From 12d1f74b6a2d55124d6a93f00b2cd985280f56df Mon Sep 17 00:00:00 2001
From: Russell King <rmk+kernel@arm.linux.org.uk>
Date: Mon, 26 Oct 2015 20:36:03 +0000
Subject: [PATCH 137/194] staging: etnaviv: fix use-after-free

Fix a use-after-free bug in etnaviv, caused by the new command queuing
code freeing the command buffer memory, and then deleting a contained
list node from the list of active GPU buffers.

Signed-off-by: Russell King <rmk+kernel@arm.linux.org.uk>
---
 drivers/staging/etnaviv/etnaviv_gpu.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/staging/etnaviv/etnaviv_gpu.c b/drivers/staging/etnaviv/etnaviv_gpu.c
index 8f54fa8..18eb737 100644
--- a/drivers/staging/etnaviv/etnaviv_gpu.c
+++ b/drivers/staging/etnaviv/etnaviv_gpu.c
@@ -916,8 +916,8 @@ static void retire_worker(struct work_struct *work)
 	list_for_each_entry_safe(cmdbuf, tmp, &gpu->active_cmd_list,
 				 gpu_active_list) {
 		if (fence_after_eq(fence, cmdbuf->fence)) {
-			etnaviv_gpu_cmdbuf_free(cmdbuf);
 			list_del(&cmdbuf->gpu_active_list);
+			etnaviv_gpu_cmdbuf_free(cmdbuf);
 		}
 	}
 
-- 
2.6.2

